File: examples/09_transaction.rs
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! cargo run --example 09_transaction --release

use iota_client::{signing::mnemonic::MnemonicSigner, Client, Result};
extern crate dotenv;
use dotenv::dotenv;
use std::env;

/// In this example we will send a transaction

#[tokio::main]
async fn main() -> Result<()> {
    let client = Client::builder()
        .with_node("http://localhost:14265")?
        .with_node_sync_disabled()
        .finish()
        .await?;

    // This example uses dotenv, which is not safe for use in production
    // Configure your own seed in ".env". Since the output amount cannot be zero, the seed must contain non-zero balance
    dotenv().ok();
    let signer = MnemonicSigner::try_from_mnemonic(&env::var("NONSECURE_USE_OF_DEVELOPMENT_MNEMONIC1").unwrap())?;

    let message = client
        .message()
        .with_signer(&signer)
        // Insert the output address and amount to spent. The amount cannot be zero.
        .with_output(
            // We generate an address from our seed so that we send the funds to ourselves
            &client.get_addresses(&signer).with_range(1..2).finish().await?[0],
            1_000_000,
        )?
        .finish()
        .await?;

    println!(
        "Transaction sent: https://explorer.iota.org/devnet/message/{}",
        message.id()
    );
    Ok(())
}

File: bindings/nodejs/examples/09_transaction.js
// Copyright 2021-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// In this example we will send a transaction
async function run() {
    const { Client } = require('@iota/client');

    // client will connect to testnet by default
    const client = new Client({
        nodes: [
            {
                url: 'http://localhost:14265',
                auth: null,
                disabled: false,
            },
        ],
        localPow: true,
    });

    require('dotenv').config();
    const signer = JSON.stringify({
        Mnemonic: process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1,
    });

    try {
        // We generate an address from our seed so that we send the funds to ourselves
        const addresses = await client.generateAddresses(signer, {
            range: {
                start: 1,
                end: 2,
            },
        });

        // Insert the output address and amount to spend. The amount cannot be zero.
        const message = await client.generateMessage(signer, {
            output: { address: addresses[0], amount: 1000000 },
        });
        console.log('Message: ', message, '\n');

        // Send transaction
        const messageId = await client.postMessage(message);

        // TODO: link doesn't work (Not found), same goes for the rust example (06_simple_message.rs)
        console.log(
            `Transaction sent: https://explorer.iota.org/devnet/message/${messageId}`,
        );
    } catch (error) {
        console.log('Error: ', error);
    }
}

run().then(() => process.exit());
